File: communicationAPI/src/main/java/eu/cybergeiger/communication/CommunicationHelper.java
package eu.cybergeiger.communication;

//import java.util.concurrent.TimeoutException;

import java.io.IOException;

/**
 * A helper class for sending and waiting on Messages.
 * TODO should this only be used for Testing?
 */
public class CommunicationHelper {

  /**
   * Interface to denote a MessageFilter.
   */
  public static interface MessageFilter {
    boolean filter(Message msg);
  }

  private static class Listener implements PluginListener {

    private MessageFilter filter;
    private LocalApi api;
    private final Object obj = new Object();
    private Message msg = null;


    public Listener(LocalApi api, MessageFilter filter) {
      if (api == null) {
        throw new NullPointerException("api may not be null");
      }
      if (filter == null) {
        throw new NullPointerException("message filter may not be null");
      }
      this.filter = filter;
      this.api = api;
      this.api.registerListener(new MessageType[]{MessageType.ALL_EVENTS}, this);
    }

    @Override
    public void pluginEvent(GeigerUrl url, Message msg) {
      if (filter.filter(msg)) {
        this.msg = msg;
        synchronized (obj) {
          obj.notifyAll();
        }
      }
    }

    @Override
    public byte[] toByteArray() {
      // TODO
      return new byte[0];
    }

    public void dispose() {
      api.deregisterListener(new MessageType[]{MessageType.ALL_EVENTS}, this);
    }

    public Message waitForResult(long timeout) throws IOException {
      //long starttime = (new Time()).getTime();
      long starttime = ch.fhnw.geiger.totalcross.System.currentTimeMillis();
      while (msg == null && (timeout < 0
          || (ch.fhnw.geiger.totalcross.System.currentTimeMillis() - starttime < timeout))) {
        try {
          synchronized (obj) {
            obj.wait(100);
          }
        } catch (InterruptedException e) {
          //safe to ignore
        }
      }
      if (msg == null) {
        throw new IOException("timeout reached");
      }
      return msg;
    }
  }

  /**
   * <p>Sends a message and waits for the first message matching the provided message filter.</p>
   *
   * @param api    the API to be used as communication endpoint
   * @param msg    the message to be sent
   * @param filter the filter matching the expected reply
   * @return the response Message
   */
  public static Message sendAndWait(LocalApi api, Message msg, MessageFilter filter)
      throws IOException {
    return sendAndWait(api, msg, filter, 10000);
  }

  /**
   * <p>Sends a message and waits for the first message matching the provided message filter.</p>
   *
   * @param api     the API to be used as communication endpoint
   * @param msg     the message to be sent
   * @param filter  the filter matching the expected reply
   * @param timeout the timeout in miliseconds (-1 for infinite)
   * @return the response Message
   */
  public static Message sendAndWait(LocalApi api, Message msg, MessageFilter filter, long timeout)
      throws IOException {
    Listener l = new Listener(api, filter);
    api.sendMessage(msg.getTargetId(), msg);
    Message result = l.waitForResult(timeout);
    l.dispose();
    return result;
  }


}

File: lib/src/communication/communication_helper.dart
library geiger_api;

import 'dart:io';

import '../../geiger_api.dart';
import 'communication_exception.dart';
import 'geiger_url.dart';
import 'message.dart';
import 'message_type.dart';
import 'plugin_listener.dart';

/// Interface to denote a message filter.
typedef MessageFilter = bool Function(Message msg);

class Listener with PluginListener {
  final MessageFilter filter;
  final GeigerApi api;
  final Object obj = Object();
  Message? msg;

  Listener(this.api, this.filter);

  Future<void> register() async {
    await api.registerListener(<MessageType>[MessageType.ALL_EVENTS], this);
  }

  @override
  void pluginEvent(GeigerUrl url, Message msg) {
    if (filter(msg)) {
      this.msg = msg;
      // synchronized(obj, {
      // obj.notifyAll();
      // });
    }
  }

  void dispose() {
    api.deregisterListener([MessageType.ALL_EVENTS], this);
  }

  Message waitForResult(int timeout) {
    var startTime = DateTime.now().millisecondsSinceEpoch;
    while ((msg == null) &&
        ((timeout < 0) ||
            ((DateTime.now().millisecondsSinceEpoch - startTime) < timeout))) {
      sleep(Duration(milliseconds: 100));
    }
    var message = msg;
    if (message == null) {
      throw CommunicationException('timeout reached');
    }
    return message;
  }
}

/// A helper class for sending and waiting on [Message]s.
/// TODO should this only be used for Testing?
class CommunicationHelper {
  /// Sends [msg] and waits for the first message matching the provided [filter].
  ///
  /// Will communication using the provided [api] and waits maximum [timeout]
  /// milliseconds. Specify `-1` to remove any time limit.
  ///
  /// Throws [CommunicationException] if communication with master fails
  static Future<Message> sendAndWait(
      GeigerApi api, Message msg, MessageFilter filter,
      [int timeout = 10000]) async {
    var l = Listener(api, filter);
    await l.register();
    await api.sendMessage(msg.targetId!, msg);
    var result = l.waitForResult(timeout);
    l.dispose();
    return result;
  }
}
